### PR TITLE
Fix inlined number params in native queries

### LINKED ISSUE #29690: Filter values no longer work inside single quoted strings
**Describe the bug**

After upgrading to version v0.46 queries that worked before are returning SQL errors that the shouldn't.

**To Reproduce**
A question with a simple SQL query:

`select replace(replace(phone, ' ', ''), '-', '') from phone_numbers where active=true and preferred=true and phone LIKE '+1 312%' limit 1;`

now returns the following error `error: The column index is out of range: 1, number of columns: 0`

note there have been no schema changes. When I take the "preview" SQL and paste it directly into psql I get a result.

**Expected behavior**

After upgrading the results should remain the same.

**Screenshots/videos**

image

**Information about your Metabase Installation:**

```
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/111.0",
    "vendor": ""
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.6+10-post-Debian-1bpo91",
    "java.vendor": "Debian",
    "java.vendor.url": "https://tracker.debian.org/openjdk-11",
    "java.version": "11.0.6",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.6+10-post-Debian-1bpo91",
    "os.name": "Linux",
    "os.version": "4.9.0-11-amd64",
    "user.language": "en",
    "user.timezone": "Etc/UTC"
  },
  "metabase-info": {
    "databases": [
      "postgres",
      "googleanalytics"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "11.5 (Debian 11.5-3.pgdg90+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.5.0"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2023-02-19",
      "tag": "v0.45.3",
      "branch": "release-x.45.x",
      "hash": "070f57b"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**
How severe an issue is this bug to you? Is this annoying, blocking some users, blocking an upgrade or blocking your usage of Metabase entirely?
Note: the more honest and specific you are here the more we will take you seriously.

I rolled back to v0.45.3 in the meantime

**Additional context**

### LINKED ISSUE #30103: [Snowflake] Syntax error: unexpected "?" when using filters inside window frames, or inside INTERVAL syntax
**Describe the bug**

When running a query with a filter inside a window frame or an INTERVAL syntax, Metabase crash since v0.46. It worked fine in v0.45.

We are connected to a managed Snowflake database on AWS.

Here is a minimal reproduction with the window frame:
```
SELECT sum(1) over (ORDER BY foo ROWS BETWEEN {{ my_filter }} PRECEDING AND CURRENT ROW)
FROM bar
```

Here is a minimal reproduction with the INTERVAL syntax:
```
SELECT CURRENT_DATE - INTERVAL '{{n}} days'
```

image

Here is the error message:
```
SQL compilation error: syntax error line 1 at position 46 unexpected '?'.
```

It looks like an issue with prepared statements. I was able to reproduce it on a local Metabase instance.

Forcing the usage of client-side prepared statements by adding `prepareThreshold=0` to the JCBD connection arguments does not solve the issue.

**Logs**

Javascript logs are a spam of:
```
error SQL compilation error: syntax error line 1 at position 46 unexpected '?'.
StructuredQuery.ts:352 

Warning: can't clean query without metadata!
value @ StructuredQuery.ts:352
```
image

Troubleshooting logs:
```
[7b98443a-ae5a-455b-b215-3decd57cde77] 2023-04-14T11:19:54+02:00 ERROR metabase.query-processor.middleware.catch-exceptions Error processing query: SQL compilation error:
syntax error line 1 at position 46 unexpected '?'.
{:database_id 2,
 :started_at #t "2023-04-14T11:19:53.712779+02:00[Europe/Paris]",
 :via
 [{:status :failed,
   :class clojure.lang.ExceptionInfo,
   :error "Error executing query: SQL compilation error:\nsyntax error line 1 at position 46 unexpected '?'.",
   :stacktrace
   ["--> driver.sql_jdbc.execute$execute_reducible_query$fn__71786.invoke(execute.clj:505)"
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:502)"
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:489)"
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:497)"
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:489)"
    "driver.sql_jdbc$fn__105090.invokeStatic(sql_jdbc.clj:63)"
    "driver.sql_jdbc$fn__105090.invoke(sql_jdbc.clj:61)"
    "query_processor.context$executef.invokeStatic(context.clj:60)"
    "query_processor.context$executef.invoke(context.clj:49)"
    "query_processor.context.default$default_runf.invokeStatic(default.clj:68)"
    "query_processor.context.default$default_runf.invoke(default.clj:66)"
    "query_processor.context$runf.invokeStatic(context.clj:46)"
    "query_processor.context$runf.invoke(context.clj:40)"
    "query_processor.reducible$identity_qp.invokeStatic(reducible.clj:15)"
    "query_processor.reducible$identity_qp.invoke(reducible.clj:12)"
    "query_processor.middleware.cache$maybe_return_cached_results$maybe_return_cached_results_STAR___69171.invoke(cache.clj:224)"
    "query_processor.middleware.permissions$check_query_permissions$fn__64643.invoke(permissions.clj:126)"
    "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__68062.invoke(mbql_to_native.clj:24)"
    "query_processor$fn__70690$combined_post_process__70695$combined_post_process_STAR___70696.invoke(query_processor.clj:243)"
    "query_processor$fn__70690$combined_pre_process__70691$combined_pre_process_STAR___70692.invoke(query_processor.clj:240)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__69082$fn__69087.invoke(resolve_database_and_driver.clj:36)"
    "driver$do_with_driver.invokeStatic(driver.clj:90)"
    "driver$do_with_driver.invoke(driver.clj:86)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__69082.invoke(resolve_database_and_driver.clj:35)"
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__64952.invoke(fetch_source_query.clj:310)"
    "query_processor.middleware.store$initialize_store$fn__65130$fn__65131.invoke(store.clj:12)"
    "query_processor.store$do_with_store.invokeStatic(store.clj:47)"
    "query_processor.store$do_with_store.invoke(store.clj:41)"
    "query_processor.middleware.store$initialize_store$fn__65130.invoke(store.clj:11)"
    "query_processor.middleware.normalize_query$normalize$fn__69371.invoke(normalize_query.clj:25)"
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__66308.invoke(constraints.clj:54)"
    "query_processor.middleware.process_userland_query$process_userland_query$fn__69307.invoke(process_userland_query.clj:150)"
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__69684.invoke(catch_exceptions.clj:171)"
    "query_processor.reducible$async_qp$qp_STAR___59455$thunk__59457.invoke(reducible.clj:103)"
    "query_processor.reducible$async_qp$qp_STAR___59455$fn__59459.invoke(reducible.clj:108)"],
   :error_type :invalid-query,
   :ex-data
   {:driver :snowflake,
    :sql
    ["-- Metabase:: userID: 1 queryType: native queryHash: 5595076d0ee9e6b2e2407be4f1ed08ed89b19f74cb4c8022f23e02e8e110922c"
     "SELECT"
     "  sum(1) over ("
     "    ORDER BY"
     "      foo ROWS BETWEEN ? PRECEDING"
     "      AND CURRENT ROW"
     "  )"
     "FROM"
     "  bar"],
    :params [6],
    :type :invalid-query}}],
 :state "42000",
 :error_type :invalid-query,
 :json_query
 {:constraints {:max-results 10000, :max-results-bare-rows 2000},
  :type :native,
  :middleware {:js-int-to-string? true, :ignore-cached-results? false, :process-viz-settings? false},
  :native
  {:template-tags
   {"my_filter"
    {:id "203c3bd1-65d7-ac17-cc9d-391d4db57e74", :name "my_filter", :display-name "My filter", :type :number}},
   :query "SELECT sum(1) over (ORDER BY foo ROWS BETWEEN {{ my_filter }} PRECEDING AND CURRENT ROW)\nFROM bar"},
  :database 2,
  :parameters
  [{:id "203c3bd1-65d7-ac17-cc9d-391d4db57e74",
    :type "number/=",
    :value [6],
    :target ["variable" ["template-tag" "my_filter"]]}],
  :async? true,
  :cache-ttl nil},
 :status :failed,
 :class net.snowflake.client.jdbc.SnowflakeSQLException,
 :stacktrace
 ["net.snowflake.client.jdbc.SnowflakeUtil.checkErrorAndThrowExceptionSub(SnowflakeUtil.java:127)"
  "net.snowflake.client.jdbc.SnowflakeUtil.checkErrorAndThrowException(SnowflakeUtil.java:67)"
  "net.snowflake.client.core.StmtUtil.pollForOutput(StmtUtil.java:456)"
  "net.snowflake.client.core.StmtUtil.execute(StmtUtil.java:366)"
  "net.snowflake.client.core.SFStatement.executeHelper(SFStatement.java:464)"
  "net.snowflake.client.core.SFStatement.executeQueryInternal(SFStatement.java:192)"
  "net.snowflake.client.core.SFStatement.executeQuery(SFStatement.java:129)"
  "net.snowflake.client.core.SFStatement.execute(SFStatement.java:742)"
  "net.snowflake.client.core.SFStatement.execute(SFStatement.java:652)"
  "net.snowflake.client.jdbc.SnowflakeStatementV1.executeQueryInternal(SnowflakeStatementV1.java:236)"
  "net.snowflake.client.jdbc.SnowflakePreparedStatementV1.executeQuery(SnowflakePreparedStatementV1.java:118)"
  "com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:1471)"
  "--> driver.sql_jdbc.execute$fn__71709.invokeStatic(execute.clj:380)"
  "driver.sql_jdbc.execute$fn__71709.invoke(execute.clj:378)"
  "driver.sql_jdbc.execute$execute_statement_or_prepared_statement_BANG_.invokeStatic(execute.clj:393)"
  "driver.sql_jdbc.execute$execute_statement_or_prepared_statement_BANG_.invoke(execute.clj:389)"
  "driver.sql_jdbc.execute$execute_reducible_query$fn__71786.invoke(execute.clj:503)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:502)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:489)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:497)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:489)"
  "driver.sql_jdbc$fn__105090.invokeStatic(sql_jdbc.clj:63)"
  "driver.sql_jdbc$fn__105090.invoke(sql_jdbc.clj:61)"
  "query_processor.context$executef.invokeStatic(context.clj:60)"
  "query_processor.context$executef.invoke(context.clj:49)"
  "query_processor.context.default$default_runf.invokeStatic(default.clj:68)"
  "query_processor.context.default$default_runf.invoke(default.clj:66)"
  "query_processor.context$runf.invokeStatic(context.clj:46)"
  "query_processor.context$runf.invoke(context.clj:40)"
  "query_processor.reducible$identity_qp.invokeStatic(reducible.clj:15)"
  "query_processor.reducible$identity_qp.invoke(reducible.clj:12)"
  "query_processor.middleware.cache$maybe_return_cached_results$maybe_return_cached_results_STAR___69171.invoke(cache.clj:224)"
  "query_processor.middleware.permissions$check_query_permissions$fn__64643.invoke(permissions.clj:126)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__68062.invoke(mbql_to_native.clj:24)"
  "query_processor$fn__70690$combined_post_process__70695$combined_post_process_STAR___70696.invoke(query_processor.clj:243)"
  "query_processor$fn__70690$combined_pre_process__70691$combined_pre_process_STAR___70692.invoke(query_processor.clj:240)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__69082$fn__69087.invoke(resolve_database_and_driver.clj:36)"
  "driver$do_with_driver.invokeStatic(driver.clj:90)"
  "driver$do_with_driver.invoke(driver.clj:86)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__69082.invoke(resolve_database_and_driver.clj:35)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__64952.invoke(fetch_source_query.clj:310)"
  "query_processor.middleware.store$initialize_store$fn__65130$fn__65131.invoke(store.clj:12)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:47)"
  "query_processor.store$do_with_store.invoke(store.clj:41)"
  "query_processor.middleware.store$initialize_store$fn__65130.invoke(store.clj:11)"
  "query_processor.middleware.normalize_query$normalize$fn__69371.invoke(normalize_query.clj:25)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__66308.invoke(constraints.clj:54)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__69307.invoke(process_userland_query.clj:150)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__69684.invoke(catch_exceptions.clj:171)"
  "query_processor.reducible$async_qp$qp_STAR___59455$thunk__59457.invoke(reducible.clj:103)"
  "query_processor.reducible$async_qp$qp_STAR___59455$fn__59459.invoke(reducible.clj:108)"],
 :card_id 1,
 :context :question,
 :error "SQL compilation error:\nsyntax error line 1 at position 46 unexpected '?'.",
 :row_count 0,
 :running_time 0,
 :data {:rows [], :cols []}}
```

**To Reproduce**

1. Connect to a Snowflake database
2. Run the query mentioned above

**Expected behavior**
The filter should get replaced properly. Instead, it isn't replaced, and the query fails.

**Screenshots/videos**
If applicable, add screenshots/videos to help explain your problem.

**Information about your Metabase Installation:**

Diagnostic information (local environment):
```
{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.18+0",
    "java.vendor": "Homebrew",
    "java.vendor.url": "https://github.com/Homebrew/homebrew-core/issues",
    "java.version": "11.0.18",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.18+0",
    "os.name": "Mac OS X",
    "os.version": "13.1",
    "user.language": "en",
    "user.timezone": "Europe/Paris"
  },
  "metabase-info": {
    "databases": [
      "snowflake"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "application-database-details": {
      "database": {
        "name": "H2",
        "version": "2.1.212 (2022-04-09)"
      },
      "jdbc-driver": {
        "name": "H2 JDBC Driver",
        "version": "2.1.212 (2022-04-09)"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2023-04-05",
      "tag": "v0.46.1",
      "branch": "release-x.46.x",
      "hash": "12a6e1d"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

This also happens on our production environment, with a Postgres internal database.

**Severity**
High. Around 15% of our dashboards are affected, and there's no way to fix these queries.

Since Metabase does not officially support downgrading (and we have no backup of our PG database before the migration), we are stuck with broken dashboards.

**Additional context**

We also noticed we can't access the "Data Model" tab anymore, it straight up fails by throwing a huge HTML log.

### PR DESCRIPTION
**Pull request by john-metabase**
Native query parameters like in `...INTERVAL '{{days}} days'` were broken due to generating incorrect inlining forms for the HoneySQL version in-use. This PR fixes number parameter inlining and adds a `sql.qp/with-driver-honey-sql-version` macro and updates places where we previously bound `*honey-sql-version*` based on a driver.

Resolves #29690, resolves #30103

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30265)

### CODE REVIEW

**Comment by john-metabase** [comment]
This should cover all cases where number params are used in interval expressions, window function frame clauses, etc. 👍

File: src/metabase/driver/sql/query_processor.clj [hunk 1]
```diff
@@ -106,6 +106,12 @@
     2
     1))
 
+(defmacro with-driver-honey-sql-version
+  "Evaluates body with the appropriate driver version of honey sql bound"
+  [driver & body]
```
Comment (metamben) [inline src/metabase/driver/sql/query_processor.clj:111]: What about adding a `:style/indent` annotation??

**Review by metamben — APPROVED** [review]
LGTM

File: src/metabase/driver/sql/query_processor.clj (hunk 1 above)
Comment (john-metabase) [reply src/metabase/driver/sql/query_processor.clj:111]: Will do 👍

**Pull request merged**